### PR TITLE
Correctly model JSON repr of complex nested records.

### LINKED ISSUE #2354: Support (non-legacy) SQL nested data types
`STRUCT<t>` and `ARRAY<t>` seem to be the types.

See #2342. h/t to @c0b for filing.

---

From #2585: [Sample app that reproduces this](https://github.com/GoogleCloudPlatform/python-docs-samples/blob/176f161cae3d4f01b01456a6e403faf6645414a1/bigquery/cloud-client/simple_app.py)

Stacktrace:

``` python
Traceback (most recent call last):
  File "simple_app.py", line 57, in <module>
    query_shakespeare()
  File "simple_app.py", line 47, in query_shakespeare
    page_token=page_token)
  File "lib/python2.7/site-packages/google/cloud/bigquery/query.py", line 401, in fetch_data
    rows_data = _rows_from_json(response.get('rows', ()), self.schema)
  File "lib/python2.7/site-packages/google/cloud/bigquery/_helpers.py", line 98, in _rows_from_json
    for item in cell['v']])
  File "lib/python2.7/site-packages/google/cloud/bigquery/_helpers.py", line 61, in _record_from_json
    for subfield, cell in zip(field.fields, value['f']):
KeyError: 'f'
```

---

Stepping into the debugger:

``` python
ipdb> value
{u'v': {u'f': [{u'v': u'hamlet'}, {u'v': u'5318'}]}}
ipdb> field.__dict__
{'field_type': u'RECORD', 'description': None, 'name': u'title', 'fields': [<google.cloud.bigquery.schema.SchemaField object at 0x7f4ba9227550>, <google.cloud.bigquery.schema.SchemaField object at 0x7f4ba9227690>], 'mode': u'REPEATED'}
```

(Added by @dhermes)

### PR DESCRIPTION
**Pull request by tseaver**
Closes #2354.

### CODE REVIEW

File: bigquery/unit_tests/test__helpers.py [hunk 1]
```diff
@@ -186,7 +186,7 @@ def test_w_scalar_subfield(self):
     def test_w_repeated_subfield(self):
         subfield = _Field('REPEATED', 'color', 'STRING')
         field = _Field('REQUIRED', fields=[subfield])
-        value = {'f': [{'v': ['red', 'yellow', 'blue']}]}
+        value = {'f': [{'v': [{'v': 'red'}, {'v': 'yellow'}, {'v': 'blue'}]}]}
```
Comment (daspecster) [inline bigquery/unit_tests/test__helpers.py]: Was this not representative of a real scenario before or are these two separate cases?

Comment (tseaver) [inline bigquery/unit_tests/test__helpers.py]: It was an incorrect guess at how the nested array was represented (the docs have no examples of the mapping).

**Comment by daspecster** [comment]
AFAICT this looks good. The nesting is a bit mesmerizing though. I would think a system test might be helpful? Should we make an issue to add a system test after this PR?

File: bigquery/google/cloud/bigquery/_helpers.py [hunk 2]
```diff
@@ -58,10 +58,11 @@ def _record_from_json(value, field):
     """Coerce 'value' to a mapping, if set or not nullable."""
     if _not_null(value, field):
         record = {}
-        for subfield, cell in zip(field.fields, value['f']):
+        record_iter = zip(field.fields, value['f'])
+        for subfield, cell in record_iter:
             converter = _CELLDATA_FROM_JSON[subfield.field_type]
-            if field.mode == 'REPEATED':
-                value = [converter(item, subfield) for item in cell['v']]
+            if subfield.mode == 'REPEATED':
```
Comment (dhermes) [inline bigquery/google/cloud/bigquery/_helpers.py]: :+1: It's pretty satisfying that this is the only change needed

**Comment by tseaver** [comment]
@daspecster Good point -- I will add a system test using the examples I generated to test the stuff manually.

**Pull request merged**